### ISSUE TITLE
Feature set description and possible label leakage

### ISSUE THREAD

**Issue by zjost** [YingtongDou/CARE-GNN#12]
Hello.  I'm trying to understand the node features.  Specifically, the referenced [paper](https://arxiv.org/pdf/2005.10150.pdf) that explains the features in Table 1 says that among the features are: "The ratio and mean of helpful and unhelpful votes".  However, my the paper says the fraud label itself is a result of the following: "Then, a user is benign if the proportion of helpful votes is higher than 0.7, and fraudulent if it is below 0.3".  This seems like a label leak, where the label is defined as a simple transformation on one or two of the features.  

My attempt of assigning meaning to the features by reading the table and playing with the data is listed below.  The format is `{column index}: {Description}; {min}, {max}, {mean}`
- 0: Number of rated products; 1, 483, 3.3
- 1: Length of user name; 0, 49, 14
- 2-6: Number of each rating level (1-5) given by user; 0, 149, 0.33
- 7-11: Proportion of each rating level; 0, 1, 0.19
- 12: Ratio of negative (1-2) ratings; 0, 1, 0.14
- 13: Ratio of positive (4-5) ratings; 0, 1, 0.77
- 14: Entropy of ratings; -9E-6, 1.6, 0.2
- 15: median review value; 1, 5, 4.1
- 16: max review value; 1, 5, 4.3
- 17: min review value; 1, 5, 3.7
- 18: avg review value; 1, 5, 4.1
- 19: number of unhelpful votes?
- 20: number of helpful votes?
- 21: time entropy; 0, 2.4, 0.26
- 22: same date indicator; 0, 1, 0.64
- 23: feedback summary length; 1, 128, 27
- 24: review text sentiment; -1, 1, 0.8

If that is correct, then 19 and 20 would be used to define the label and therefore represent a leak.  

When I simply use this column as the model score, I get a ROC AUC of 0.88-0.89, which seems to approximately match the performance [reported in the DGL implementation](https://github.com/dmlc/dgl/tree/master/examples/pytorch/caregnn).  Furthermore, if I simply train a Random Forest model on the full set of features (i.e., no graph structure used), I get a ROC AUC of 0.97, which is far higher than even CARE-GNN.

Can you please help understand these questions about features, label and performance?

**Comment by YingtongDou** [comment]
Hi @zjost,

Thanks for your interest in my paper and code! I also found the label leakage problem when I was replicating the features from the referenced paper. I have sent an email to its authors but got no response.

In my paper, I have discarded all features related to helpful votes and reduced the feature dimension of Amazon data from 36 to 25 to avoid label leakage. I have uploaded the amazon data preprocessing [code](https://github.com/YingtongDou/CARE-GNN/blob/master/amazon_preprocess.py) to this repo. You can find how I calculate each feature in Line 230 and obtain the 25-dimensional feature vectors in Line 319.

Since this paper discussed the camouflage issue of fraudsters when applying GNNs to a noisy graph, the performance of CARE-GNN is not necessarily better than Random Forest. Besides, the tree-based model has been [proven](https://arxiv.org/pdf/2207.08815.pdf) to handle hand-crafted numerical features better than deep learning models. 

If you check [this leaderboard](https://paperswithcode.com/sota/node-classification-on-amazon-fraud), some follow-up works improve the ROC-AUC on Amazon to 0.97 which seems to be the optimal performance on this dataset.

To summarize, the Amazon data is not a good dataset for fraud detection since its task is too easy but there were not too many open-source fraud datasets when I was working on this project. As I mentioned above, my paper and the follow-up works didn't aim to make a simple problem complicated; the proposed solutions can be used in more appropriate datasets where tree-based models are suboptimal and the graph structure plays a more vital role.

Please let me know if you have any other questions.

**Comment by zjost** [comment]
Hi @YingtongDou , thanks so much for your response and for taking the time to upload additional code to show your work.  I think the problem is [line 321](https://github.com/YingtongDou/CARE-GNN/blob/master/amazon_preprocess.py#L321): `new_features = np.hstack([new_features, features[:, 30:]])`.  This should be `new_features = np.hstack([new_features, features[:, 31:]])` because the 30th column is the minimum number of unhelpful votes, as defined [here](https://github.com/YingtongDou/CARE-GNN/blob/master/amazon_preprocess.py#L121).  I think there should only be 24 features, not 25.

If you calculate the ROC AUC score for each feature column with the label, you'll see that this column (column 19) corresponds to the highest AUC.

**Comment by YingtongDou** [comment]
That makes sense. I think I made a mistake with the index when filtering the vote-related features. Thanks for letting me know about this problem. I will update the code and readme to reflect this issue.
